packages/components/src/components/ui/PanelWithToolbar/index.tsx

```diff
@@ -0,0 +1,57 @@
+import { ReactNode, useCallback, useState } from "react";
+import { ToolbarItem } from "./ToolbarItem.js";
+
+type OpenModal<ModalNames extends string[]> = (
+  modalName: ModalNames[number]
+) => void;
+
+type Props<ModalNames extends string[]> = {
+  renderToolbar(props: { openModal: OpenModal<ModalNames> }): ReactNode;
+  renderBody(): ReactNode;
+  renderModal?: (modalName: ModalNames[number]) =>
+    | {
+        title: string;
+        body: ReactNode;
+      }
+    | undefined; // technically undefined shouldn't be allowed, but that would cause too many type issues
+};
+
+export function PanelWithToolbar<const ModalNames extends string[]>({
```

Comment:
This props stuff seems a bit gnarly, I assume you have decent reasons though.

Reply:
Not in this PR (turned out it's pretty complicated to pass these generics through LeftPlaygroundPanel and SquigglePlayground), but I want to make use of it eventually.

If we discover later that we don't want any more panels like `exports`, we can simplify it and use plain strings.

Reply:
Makes sense